src/proposer/finalization_reward_logic.h

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2019 The Unit-e developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://opensource.org/licenses/MIT.
+
+#ifndef UNITE_PROPOSER_FINALIZATION_REWARD_LOGIC_H
+#define UNITE_PROPOSER_FINALIZATION_REWARD_LOGIC_H
+
+#include <blockchain/blockchain_behavior.h>
+#include <blockdb.h>
+#include <dependency.h>
+#include <finalization/state_repository.h>
+
+#include <memory>
+#include <vector>
+
+namespace proposer {
+
+class MissingBlockError : public std::runtime_error {
+ public:
+  const CBlockIndex &missed_index;
+  explicit MissingBlockError(const CBlockIndex &index) noexcept
+      : std::runtime_error(strprintf("Cannot load block=%s", index.GetBlockHash().GetHex())),
+        missed_index(index) {}
+};
+
+class FinalizationRewardLogic {
+ public:
+  //! \brief Calculate proposer finalization rewards
+  //!
+  //! The first block of every epoch (i.e. previous_block is a checkpoint) must
+  //! contain finalization rewards for the block proposers of the previous
+  //! epoch. The number of finalization reward outputs equals the epoch length.
```

Comment:
> The number of finalization reward outputs equals the epoch length

As indicated in #418, we should also consider non-optimal finalization, where the dynasty is several epochs long.

Reply:
We decided to pay out rewards after each checkpoint but make rewards mature only after finalization. That means that the number of reward outputs always equals the epoch length.